digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -506,7 +519,150 @@ trait LaneOperations {
+    val roadLinksWithRoadAddressInfo = viiteInformation.filter(_.attributes.contains("VIITE_ROAD_NUMBER")) ++ vkmInformation
+
+
+    val upToDateLaneChanges = upToDateLanes.flatMap{ upToDate =>
+      val roadLink = roadLinksWithRoadAddressInfo.find(_.linkId == upToDate.linkId)
+      val relevantHistory = historyLanes.find(history => history.newId == upToDate.id)
+
+      relevantHistory match {
+        case Some(history) =>
+          val uptoDateLastModification = historyLanes.filter(historyLane =>
+            history.newId == historyLane.oldId && historyLane.newId == 0 &&
+              (historyLane.historyCreatedDate.isAfter(history.historyCreatedDate) || historyLane.historyCreatedDate.isEqual(history.historyCreatedDate)))
+
+          if (uptoDateLastModification.nonEmpty && upToDate.laneCode != uptoDateLastModification.minBy(_.historyCreatedDate.getMillis).laneCode)
+            Some(LaneChange(upToDate, Some(historyLaneToPersistedLane(uptoDateLastModification.minBy(_.historyCreatedDate.getMillis))), LaneChangeType.LaneCodeTransfer, roadLink))
+
+          else if (historyLanes.count(historyLane => historyLane.newId != 0 && historyLane.oldId == history.oldId) == 2)
+            Some(LaneChange(upToDate, Some(historyLaneToPersistedLane(history)), LaneChangeType.Divided, roadLink))
+
+          else if (upToDate.endMeasure - upToDate.startMeasure > history.endMeasure - history.startMeasure)
+            Some(LaneChange(upToDate, Some(historyLaneToPersistedLane(history)), LaneChangeType.Lengthened, roadLink))
+
+          else if(upToDate.endMeasure - upToDate.startMeasure < history.endMeasure - history.startMeasure)
+            Some(LaneChange(upToDate, Some(historyLaneToPersistedLane(history)), LaneChangeType.Shortened, roadLink))
+
+          else
+            None
+
+        case None if upToDate.modifiedDateTime.isEmpty =>
+          Some(LaneChange(upToDate, None, LaneChangeType.Add, roadLink))
+
+        case _ =>
+          val historyRelatedLanes = historyLanes.filter(_.oldId == upToDate.id)
+
+          if(historyRelatedLanes.nonEmpty){
+            val historyLane = historyLaneToPersistedLane(historyRelatedLanes.maxBy(_.historyCreatedDate.getMillis))
+
+            if (isSomePropertyDifferent(historyLane, upToDate))
+              Some(LaneChange(upToDate, Some(historyLane), LaneChangeType.AttributesChanged, roadLink))
+            else
+              None
+          }else{
+            None
+          }
+      }
+    }
+
+    val expiredLanes = historyLanes.filter(lane => lane.expired && lane.newId == 0).map{ history =>
+      val roadLink = roadLinksWithRoadAddressInfo.find(_.linkId == history.linkId)
+      LaneChange(historyLaneToPersistedLane(history), None, LaneChangeType.Expired, roadLink)
+    }
+
+    val historyLaneChanges = historyLanes.groupBy(_.oldId).flatMap{ case (_, lanes) =>
+      val roadLink = roadLinksWithRoadAddressInfo.find(_.linkId == lanes.head.linkId)
+
+      lanes.flatMap { case lane =>
```

Comment:
this case is not necessary here

Reply:
Code changed so this suggestion is now deprecated